easybuild/easyblocks/n/nvhpc.py

```diff
@@ -0,0 +1,201 @@
+##
+# Copyright 2015-2019 Bart Oldeman
+# Copyright 2016-2020 Forschungszentrum Juelich
+#
+# This file is triple-licensed under GPLv2 (see below), MIT, and
+# BSD three-clause licenses.
+#
+# This file is part of EasyBuild,
+# originally created by the HPC team of Ghent University (http://ugent.be/hpc/en),
+# with support of Ghent University (http://ugent.be/hpc),
+# the Flemish Supercomputer Centre (VSC) (https://www.vscentrum.be),
+# Flemish Research Foundation (FWO) (http://www.fwo.be/en)
+# and the Department of Economy, Science and Innovation (EWI) (http://www.ewi-vlaanderen.be/en).
+#
+# https://github.com/easybuilders/easybuild
+#
+# EasyBuild is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation v2.
+#
+# EasyBuild is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with EasyBuild.  If not, see <http://www.gnu.org/licenses/>.
+##
+"""
+EasyBuild support for installing NVIDIA HPC SDK compilers, based on the easyblock for PGI compilers
+
+@author: Bart Oldeman (McGill University, Calcul Quebec, Compute Canada)
+@author: Damian Alvarez (Forschungszentrum Juelich)
+@author: Andreas Herten (Forschungszentrum Juelich)
+"""
+import os
+import fileinput
+import re
+import stat
+import sys
+
+import easybuild.tools.environment as env
+from distutils.version import LooseVersion
+from easybuild.easyblocks.generic.packedbinary import PackedBinary
+from easybuild.framework.easyconfig import CUSTOM
+from easybuild.tools.filetools import adjust_permissions, write_file
+from easybuild.tools.run import run_cmd
+from easybuild.tools.modules import get_software_root
+
+
+# contents for siterc file to make PGI/NVHPC pick up $LIBRARY_PATH
+# cfr. https://www.pgroup.com/support/link.htm#lib_path_ldflags
+SITERC_LIBRARY_PATH = """
+# get the value of the environment variable LIBRARY_PATH
+variable LIBRARY_PATH is environment(LIBRARY_PATH);
+
+# split this value at colons, separate by -L, prepend 1st one by -L
+variable library_path is
+default($if($LIBRARY_PATH,-L$replace($LIBRARY_PATH,":", -L)));
+
+# add the -L arguments to the link line
+append LDLIBARGS=$library_path;
+
+# also include the location where libm & co live on Debian-based systems
+# cfr. https://github.com/easybuilders/easybuild-easyblocks/pull/919
+append LDLIBARGS=-L/usr/lib/x86_64-linux-gnu;
+"""
+
+
+class EB_NVHPC(PackedBinary):
+    """
+    Support for installing the NVIDIA HPC SDK (NVHPC) compilers
+    """
+
+    @staticmethod
+    def extra_options():
+        extra_vars = {
+            'default_cuda_version': ["10.1", "CUDA Version to be used as default (10.2, 11.0, ...)", CUSTOM],
```

Comment:
Does it make sense to hardcode any default version here?

How about using `None` as a default, and complaining when no specific version is specified?

At installation time, you can specify a value to be used using `eb --try-amend=default_cuda_version=10.1` for example (the error message produced when the value is still `None` could mention that, we do something like this in the [XALT easyblock](https://github.com/easybuilders/easybuild-easyblocks/blob/5e2a7577f2ddcc04b0364b577b9d84b520842540/easybuild/easyblocks/x/xalt.py#L71-L81)).

The problem with using a default version here is that it'll go irrelevant in a heart beat (well, ok, months or years) since CUDA is updated regularly.

Reply:
Actually there a better option than throwing an error: use the CUDA version of the loaded module (since you include `CUDA` as a dep)?

You can obtain it via `get_software_version('CUDA')` (could be `None` if `CUDA` is not loaded/not included as a dep)

Reply:
Let's answer your second suggestion first.

*I* do include the CUDA module as a dependency, but it is not strictly necessary. That's what `module_add_cuda` is about. NVHPC is shipped with a CUDA toolkit installation which actually the NVHPC owners would like you to use – as it is tested with NVHPC. For Jülich, I don't like this, because we _do_ have a reference system CUDA installation which should be the one in use, and not a shadow-version introduced with NVHPC. That said, I don't want to keep anyone from doing things differently (and for us, opting to do so in the future as well).

Now, coming back to your first suggestion: Having `None` is a good idea. It can not only be overridden by the command line flag `--try-amend` but still via the easyconfig, right? I'd like to have as much in the easyconfig as possible to make it documented concisely (and so I remember next time I install a NVHPC version 😀 ).

I'd implement the following sketch then:

* If `default_cuda_version` is set (which ever way), use this. If it is not,
* if `get_software_version('CUDA')` returns something, use this. If it does not,
* error out, as in XALT.

Does that sound good? I'll dive into how to do that then.

Reply:
I've implemented it accordingly in 25c798a7f5126769d09ba9eab3aa3f12952b48b3; I tested as much as I can and it seems to work.

One thing to note: In the easyconfig file (yet to be updated), `default_cuda_version = None` needs to be specified. Otherwise, EasyBuild would not allow overwriting it from the command line with the following error:

```
ERROR: Unknown easyconfig parameter: default_cuda_version (known: ['allow_prepend_abs_path', 'allow_system_deps', 'bitbucket_account', 'block', 'builddependencies', 'buildininstalldir', 'buildopts', 'buildstats', 'checksums', 'cleanupoldbuild', 'cleanupoldinstall', 'configopts', 'cuda_compute_capabilities', 'dependencies', 'deprecated', 'description', 'docpaths', 'docurls', 'dontcreateinstalldir', 'easyblock', 'easybuild_version', 'enhance_sanity_check', 'examples', 'exts_classmap', 'exts_default_options', 'exts_defaultclass', 'exts_download_dep_fail', 'exts_filter', 'exts_list', 'fix_perl_shebang_for', 'fix_python_shebang_for', 'github_account', 'group', 'hidden', 'hiddendependencies', 'homepage', 'include_modpath_extensions', 'installopts', 'keeppreviousinstall', 'keepsymlinks', 'key', 'license_file', 'license_server', 'license_server_port', 'maxparallel', 'modaliases', 'modaltsoftname', 'moddependpaths', 'modextrapaths', 'modextravars', 'modloadmsg', 'modluafooter', 'modtclfooter', 'module_depends_on', 'moduleclass', 'moduleforceunload', 'moduleloadnoconflict', 'multi_deps', 'multi_deps_load_default', 'name', 'onlytcmod', 'osdependencies', 'parallel', 'patches', 'postinstallcmds', 'prebuildopts', 'preconfigopts', 'preinstallopts', 'pretestopts', 'recursive_module_unload', 'runtest', 'sanity_check_commands', 'sanity_check_paths', 'site_contacts', 'skip', 'skipsteps', 'software_license', 'software_license_urls', 'source_urls', 'sources', 'start_dir', 'stop', 'testopts', 'tests', 'toolchain', 'toolchainopts', 'unpack_options', 'unwanted_env_vars', 'upstream_contacts', 'usage', 'version', 'versionprefix', 'versionsuffix', 'whatis'])
```

This is the intended behaviour, yes?

Reply:
That lgtm. @boegel ?

Reply:
Yes, sounds good!